common/src/main/java/io/pravega/common/concurrent/ScheduledQueue.java

```diff
@@ -0,0 +1,350 @@
+                return null;
+            } 
+            if (delayedTasks.remove(delayed.getKey(), delayed.getValue())) {
+                this.itemsRemoved.incrementAndGet();
+                return delayed.getValue();
+            }
+        }
+    }
+
+    private long sleepTimeout(long timeout, long startTime, long now, Entry<FireTime, E> delayed) {
+        long sleepTimeout = timeout - (now - startTime);
+        if (delayed != null) {
+            sleepTimeout = Math.min(sleepTimeout, delayed.getKey().getTimeNanos() - now);
+        }
+        return sleepTimeout;
+    }
+    
+    /**
+     * Inserts the specified element into this delay queue.
+     *
+     * @param e the element to add
+     * @return {@code true} (as specified by {@link Collection#add})
+     * @throws NullPointerException if the specified element is null
+     */
+    @Override
+    public boolean add(E e) {
+        return offer(e);
+    }
+
+    /**
+     * Inserts the specified element into this delay queue.
+     *
+     * @param e the element to add
+     * @return {@code true}
+     * @throws NullPointerException if the specified element is null
+     */
+    @Override
+    public boolean offer(E e) {
+        long seq = itemsAdded.incrementAndGet();
+        if (!e.isDelayed()) {
+            readyTasks.add(e);
+        } else {
+            delayedTasks.put(new FireTime(e.getScheduledTimeNanos(), seq), e);
+        }
+        // This is done unconditionally even if delayed because it could be the
+        // new shortest delay in which case some thread should wake up to
+        // re-schedule its sleep.
+        blocker.release();
+        return true;
+    }
+
+    /**
+     * Inserts the specified element into this delay queue. As the queue is
+     * unbounded this method will never block.
+     *
+     * @param e the element to add
+     * @param timeout This parameter is ignored as the method never blocks
+     * @param unit This parameter is ignored as the method never blocks
+     * @return {@code true}
+     * @throws NullPointerException {@inheritDoc}
+     */
+    @Override
+    public boolean offer(E e, long timeout, TimeUnit unit) {
+        return offer(e);
+    }
+    
+    /**
+     * Inserts the specified element into this delay queue. As the queue is
+     * unbounded this method will never block.
+     *
+     * @param e the element to add
+     * @throws NullPointerException {@inheritDoc}
+     */
+    @Override
+    public void put(E e) {
+        offer(e);
+    }
+
+    /**
+     * Retrieves, but does not remove, the head of this queue, or
+     * returns {@code null} if this queue is empty. 
+     * 
+     * Unlike {@code poll}, this method returns the elements who's scheduled time has not yet arrived.
+     *
+     * @return the head of this queue, or {@code null} if this queue is empty
+     */
+    @Override
+    public E peek() {
+        val ready = readyTasks.peek();
```

Comment:
though it doesnt matter, but the behaviour of this method, because of its lockless nature is that an entry could be added to ready tasks either before this line after this method's invocation or after it has peeked here. 
so it would not give a consistent answer as compared to delayed queue. but i think its ok as long as we explain that in the javadocs.

Reply:
Added a line in the javadoc.